drivers/usb/device/usb_dc_smartbond.c

```diff
@@ -0,0 +1,1398 @@
+	}
+	if (USB->USB_TXEV_REG & BIT(1)) {
+		handle_epx_tx_ev(usb_dc_get_ep_in_state(2));
+	}
+	if (USB->USB_TXEV_REG & BIT(2)) {
+		handle_epx_tx_ev(usb_dc_get_ep_in_state(3));
+	}
+}
+
+static uint32_t check_reset_end(uint32_t alt_ev)
+{
+	if (dev_state.nfsr == NFSR_NODE_RESET) {
+		if (GET_BIT(alt_ev, USB_USB_ALTEV_REG_USB_RESET)) {
+			/*
+			 * Could be still in reset, but since USB_M_RESET is
+			 * disabled it can be also old reset state that was not
+			 * cleared yet.
+			 * If (after reading USB_ALTEV_REG register again)
+			 * bit is cleared reset state just ended.
+			 * Keep non-reset bits combined from two previous
+			 * ALTEV read and one from the next line.
+			 */
+			alt_ev = (alt_ev & ~USB_USB_ALTEV_REG_USB_RESET_Msk) |
+				 USB->USB_ALTEV_REG;
+		}
+		if (GET_BIT(alt_ev, USB_USB_ALTEV_REG_USB_RESET) == 0) {
+			USB->USB_ALTMSK_REG = USB_USB_ALTMSK_REG_USB_M_RESET_Msk |
+					      USB_USB_ALTEV_REG_USB_SD3_Msk;
+			if (dev_state.ep_state[0][0].buffer != NULL) {
+				USB->USB_MAMSK_REG |=
+					USB_USB_MAMSK_REG_USB_M_EP0_RX_Msk;
+			}
+			LOG_INF("Set operational %02x", USB->USB_MAMSK_REG);
+			set_nfsr(NFSR_NODE_OPERATIONAL);
+		}
+	}
+	return alt_ev;
+}
+
+static void handle_bus_reset(void)
+{
+	uint32_t alt_ev;
+
+	USB->USB_NFSR_REG = 0;
+	USB->USB_FAR_REG = 0x80;
+	USB->USB_ALTMSK_REG = 0;
+	USB->USB_NFSR_REG = NFSR_NODE_RESET;
+	USB->USB_TXMSK_REG = 0;
+	USB->USB_RXMSK_REG = 0;
+	set_nfsr(NFSR_NODE_RESET);
+
+	for (int i = 0; i < EP_MAX; ++i) {
+		dev_state.ep_state[1][i].buffer = NULL;
+		dev_state.ep_state[1][i].transferred = 0;
+		dev_state.ep_state[1][i].total_len = 0;
+		atomic_clear(&dev_state.ep_state[1][i].busy);
+	}
+
+	LOG_INF("send USB_DC_RESET");
+	dev_state.status_cb(USB_DC_RESET, NULL);
+	USB->USB_DMA_CTRL_REG = 0;
+
+	USB->USB_MAMSK_REG = USB_USB_MAMSK_REG_USB_M_INTR_Msk |
+			     USB_USB_MAMSK_REG_USB_M_FRAME_Msk |
+			     USB_USB_MAMSK_REG_USB_M_WARN_Msk |
+			     USB_USB_MAMSK_REG_USB_M_ALT_Msk |
+			     USB_USB_MAMSK_REG_USB_M_EP0_RX_Msk |
+			     USB_USB_MAMSK_REG_USB_M_EP0_TX_Msk;
+	USB->USB_ALTMSK_REG = USB_USB_ALTMSK_REG_USB_M_RESUME_Msk;
+	alt_ev = USB->USB_ALTEV_REG;
+	check_reset_end(alt_ev);
+}
+
+static void handle_alt_ev(void)
+{
+	struct smartbond_ep_state *ep_state;
+	uint32_t alt_ev = USB->USB_ALTEV_REG;
+
+	alt_ev = check_reset_end(alt_ev);
+	if (GET_BIT(alt_ev, USB_USB_ALTEV_REG_USB_RESET) &&
+	    dev_state.nfsr != NFSR_NODE_RESET) {
+		handle_bus_reset();
+	} else if (GET_BIT(alt_ev, USB_USB_ALTEV_REG_USB_RESUME)) {
+		if (USB->USB_NFSR_REG == NFSR_NODE_SUSPEND) {
+			set_nfsr(NFSR_NODE_OPERATIONAL);
+			if (dev_state.ep_state[0][0].buffer != NULL) {
+				USB->USB_MAMSK_REG |=
+					USB_USB_MAMSK_REG_USB_M_EP0_RX_Msk;
+			}
+			USB->USB_ALTMSK_REG = USB_USB_ALTMSK_REG_USB_M_RESET_Msk |
+					      USB_USB_ALTMSK_REG_USB_M_SD3_Msk;
+			/* Re-enable reception of endpoint with pending transfer */
+			for (int ep_num = 1; ep_num <= 3; ++ep_num) {
```

Comment:
`for (int ep_num = 1; ep_num < EP_MAX; ++ep_num) {` ?

Reply:
@NiekIlmer Please recheck your change, EP_MAX is 4. :arrow_up:

Reply:
Good catch, fixed